### ISSUE TITLE
Access to ctx.db after disconnect()

### ISSUE THREAD

**Issue by oliversalzburg** [vitaly-t/pg-promise#40]
Not quite sure about the circumstances, but I'm getting `TypeError: Cannot read property 'client' of null` in `index.js:430`.

`ctx.db` is null, probably because `disconnect()` was already called.

**Comment by vitaly-t** [comment]
Could you give me an example of the code that does that?

**Comment by oliversalzburg** [comment]
Wow, I was on 1.8.3! I just upgraded and will report back ;)

**Comment by oliversalzburg** [comment]
Well, no issues so far. I'll open another ticket if necessary :) Cheers!

**Comment by vitaly-t** [comment]
@oliversalzburg thank you for getting back with this.

I do not recall any changes that could affect the behaviour you described in the beginning with version 1.8.3, but then again, from the current version it is separated by 1,113 additions and 246 deletions, which is so many, even I can't be sure if something hasn't changed in between. We can only hope that the 100% test coverage that the library has is comprehensive enough and trustworthy.

**Comment by vitaly-t** [comment]
@oliversalzburg I was really interested to research this better, trying to understand how this could have happened. So, I went through the entire history of changes between 1.8.3 and 1.9.7, only to become absolutely sure that nothing has changed to affect that kind of behaviour you described.

Then I realised that the only way this may have happened, is through a theoretical bug within the specific library + version of the promise library that you were using. If you suddenly cannot see the issue after upgrading to 1.9.7, then I'm guessing that along the line you upgraded whatever promise library that you were using, which may in fact be the cause of the problem.

The problem itself, as I suspect it, I have 3 places in the code where I first initiate a `reject` and then immediately invoke an error event notification. The bug would have been if `reject` had immediately triggered the rejection call chain, which would have been invalid under the promise standard, and break the logic of the notification, i.e. the notification would attempt to use the connection details that was released prematurely.

Strictly speaking, in this library we do not care whether we trigger the reject first and then provide a notification or vice-versa. So, for the sake of making it more robust, and protecting against such hypothetical promise glitsh, I've just swapped the sequence in those 3 places, to provide a notification first, and only then trigger the reject.

You can see the code just checked, labelled `improved safety of the error notification.` [Line 618](https://github.com/vitaly-t/pg-promise/blob/master/lib/index.js#L618) is the theoretical culprit in your case.

I appreciate if you try to track back your recent update and let me know whether what I was suspecting is true. And if you manage to reproduce the issue by restoring your version of the promise library, then you can manually swap those lines yourself, the reject and the notification to see if that removes the issue, so then we know if the problem was indeed in the promise library as I suspected.

**Comment by vitaly-t** [comment]
Reopening the issue, to keep track of it ;)

**Comment by oliversalzburg** [comment]
I ran into this issue again, even after the upgrade. I'll read you replies and get back into the details tomorrow.

**Comment by vitaly-t** [comment]
This is probably for the better, makes it easier to track the exact issue. If you can reproduce it, and my last check-in fixes it, then we will know what the issue is and have the fix ready.

**Comment by oliversalzburg** [comment]
Well, in general, this happens when some query doesn't go as planned during a transaction (`transaction.one()` returned multiple rows). I'm still not quite sure why it happens, as it doesn't happen everytime there is an error.

The full stack is:
```
c:\Users\Oliver\core\node_modules\pg-promise\lib\index.js:619
                    client: ctx.db.client,
                                  ^
TypeError: Cannot read property 'client' of null
    at notifyReject (c:\Users\Oliver\core\node_modules\pg-promise\lib\index.js:619:35)
    at null.callback (c:\Users\Oliver\core\node_modules\pg-promise\lib\index.js:605:22)
    at Query.handleError (c:\Users\Oliver\core\node_modules\pg\lib\query.js:106:17)
    at null.<anonymous> (c:\Users\Oliver\core\node_modules\pg\lib\client.js:171:26)
    at emit (events.js:107:17)
    at Socket.<anonymous> (c:\Users\Oliver\core\node_modules\pg\lib\connection.js:107:12)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:538:20)
```

I'm currently debugging an issue in our code where a large amount of requests to the API generate race conditions with our database transactions. The failures usually end in the error above being logged and Node exiting.

Due to the nature of this issue, requiring large amounts of correctly timed queries, and the asynchronicity of the code, it is really hard to see what the underlying cause is.

If you have any suggestions regarding what I should look into, I'd be happy to investigate.

**Comment by vitaly-t** [comment]
Have you tried the updated code as per my previous posts?

What promise library are you using?

Can you give your transaction code example?

**Comment by vitaly-t** [comment]
@oliversalzburg since there was no update from you on this issue, there was another [release 1.9.8](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.9.8) with the changes that address this issue. You see all the details in the release notes.

Please let us know if the issue is gone using the new version.

**Comment by oliversalzburg** [comment]
Hey @vitaly-t. Sorry for not providing better feedback earlier. It was quite the busy weekend.

I just installed 1.9.8 and am still seeing the same error. I'm using bluebird 2.9.34.

Regarding the transaction code. The code is rather long and I would prefer not to share it publicly. I could send you a copy via email if you like.

**Comment by vitaly-t** [comment]
That's quite unexpected, because the only way that problem could be understood is through the break in the reject chain. Nothing else is logically possible there.

Ok, you can send your code to vitaly.tomilov@gmail.com, I'll have a look at it.

**Comment by oliversalzburg** [comment]
It seems to me like the rejection handler from `$connect` (line 483) is first invoked, then the callback for `query` (line 568) is.

I'll send you the code shortly.

**Comment by vitaly-t** [comment]
Yes, but you see, in 1.9.8 I swapped those two in places, so it invokes the callback first, and only then rejects, so it is impossible to lose the context for the callback. That I can't understand.

**Comment by oliversalzburg** [comment]
Do you think there is a possible relation to https://github.com/brianc/node-postgres/issues/746? Because I've been banging into that left and right. First an `error` event is handled, but then an identical message is received on the `data` stream and handled as the same error again.

**Comment by vitaly-t** [comment]
That issue related to the connection, while your error stack was pointing at the code inside the query execution, which means the connection was established successfully.

**Comment by vitaly-t** [comment]
Thank's for the code, there is so much of it, I'd like to know which piece is actually failing.

I won't be able to finish my analysis today, unfortunately, will do it tomorrow.

So for now will just leave with a tip that you seem to overuse transactions a lot there, in places where they are not needed at all, like one or more select statements. Transactions only make sense when there is at least one `INSERT` or `UPDATE` involved.

When it's just one query, you should just return the query result. And when it's a sequence of selects, [Tasks](https://github.com/vitaly-t/pg-promise#tasks) should be used.

**Comment by vitaly-t** [comment]
Also, since you are wisely tagging transactions (with name), this should help a lot analysing what's going on, if you connect [pg-monitor](https://github.com/vitaly-t/pg-monitor). Try it, see if it gives you a better picture of what is happening.

**Comment by oliversalzburg** [comment]
Good point. I'll look into reducing the number of transactions.

The point where the code is failing originates from the query in `updatePersonTags`. But this could be coincidental.

In my test scenario, I'm creating 100 person records. When I run the necessary queries in succession, everything completes successfully without errors.  

When I issue the queries to the API as fast as possible, many transactions are being generated and sent to Postgres at the same time. The client pool will fill up with connections and a lot of failures start to happen due to table locking conflicts. With the current state of the code, this is expected.

However, at some point the access to `ctx.db.client` happens and `db` is `null`, even though the same line of code was hit many times before and `ctx.db` wasn't `null` :\

I'll have a look at pg-monitor. Thanks for the suggestion.

**Comment by oliversalzburg** [comment]
Ah, I see the point of pg-monitor. We're already logging excessively and I got all the "details" at hand. But the log for a single test run is close to 8k lines. It would probably help if I tagged all transactions with a unique ID, but even then it's a lot of log.

**Comment by vitaly-t** [comment]
You can configure it easily to log only certain things, like tasks and transactions:
```javascript
monitor.attach(options, ['task', 'transact']);
```

**Comment by oliversalzburg** [comment]
I'm pretty sure this problem is not related to the actual queries. It seems like a lot of things are going wrong once we're hitting the connection limit of postgres or the connection pool size limit of node-postgres (https://github.com/coopernurse/node-pool/issues/99).

Maybe what I'm seeing in pg-promise is just a side-effect of a larger underlying issue :P

**Comment by oliversalzburg** [comment]
Nope, not related to the issue mentioned above. Working on a different issue at home, I run into this error much earlier, given invalid input for our application. I run a single `createWith` for a person and end up with the uncaught error.

I took the liberty of emailing you the log, which shows the issued queries. The API failure is expected, the TypeError obviously is not :(

Given that this is a much simpler case than before, I'll try to see if the debugger brings up any new information.

**Comment by oliversalzburg** [comment]
I've reduced the code that creates the problem to this:
```js
return self.db.tx( function tx( transaction ) {
	return Promise.all( [ foo(), fails() ] );

	function foo() {
		return succeeds()
			.then( succeeds )
			.then( fails );

		function succeeds() {
			return transaction.one( "SELECT 1 AS id" );
		}
	}

	function fails() {
		return transaction.one( "SELECT id FROM person WHERE(FALSE)" );
	}
} );
```

Interestingly, if I specify an invalid table name instead of `person`, I get a different, handled, error and execution continues just fine.

**Comment by vitaly-t** [comment]
When I execute the code above, I'm always just getting `No data returned from the query.`, a perfectly expected result.

**Comment by vitaly-t** [comment]
I am beginning to understand why this may occur, after looking at the log file that you sent me, specifically the following lines:

```
2015-09-07 19:50:00.566 [DEBUG ] (postgres.js) tx(person.createWith): rollback
2015-09-07 19:50:00.568 [ERROR ] (postgres.js) error occurred on database transaction
2015-09-07 19:50:00.568 [ERROR ] (postgres.js) SELECT id FROM "fairDay" WHERE ("fairDay"."date" = '2015-09-06T22:00:00.000Z')
2015-09-07 19:50:00.571 [DEBUG ] (postgres.js) tx(person.createWith): INSERT INTO "personTags" ("person", "tag") VALUES ('fccc3053-9666-40dd-8732-824cffc6a5c7', '44f835e9-c0b6-46f3-8ac3-865d1de24355')
```
You manage to execute requests against a transaction after the transaction executes `roolback`. This is not supposed to happen. And the only way to make this happen is by creating loose requests against the transaction context. A loose request is any request created against transaction context that's not included into the transaction logic.

Example:

```javascript
db.tx(function tx(t) {
    t.one("select login from users where(false)"); // loose request;
    return promise.reject("finished!");
})
    .then(function (data) {
        console.log("SUCCESS:", data);
    }, function (reason) {
        console.log("FAILED:", reason);
    });
```
Any loose request will break the transaction, naturally, because the transaction has no control over the execution of loose requests. It is generally a broken promise logic.

You need to check your code where you may be creating a loose request against transaction context.

In the meantime, I will consider adding detection for loose queries to the library, if it is possible.

**Comment by vitaly-t** [comment]
I have released [version 1.9.9](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.9.9) that addresses the issue with loose request detection. See my release notes.

Using this version you should be able to see the exact query that's made outside of the connection context, i.e. loose query.

**Comment by vitaly-t** [comment]
A had a moment looking through your code, to search for loose calls. I have found one obvious loose call:
```javascript
PersonManager.prototype.txCreateWith = function PersonManager$txCreateWith( transaction, data ) {
	var self = this;

	transaction.any( "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE" );
	return self.txCreate( transaction )
		.then( updateRecord );

	function updateRecord( recordId ) {
		return self.txUpdate( transaction, recordId.id, data );
	}
};
```
You cannot just call this:
```javascript
transaction.any( "SET TRANSACTION ISOLATION LEVEL SERIALIZABLE" );
```
and not provide `.then` logic for it. This would result in the exact issue you are having.

That's the only one that's obvious. You can have more, if you have a function that returns a promise object, which you call somewhere without using `.then` on the result. It's just this kind of case is too difficult to spot in someone's code. I believe the new version will point to such cases better than before.

**Comment by oliversalzburg** [comment]
At home, I replaced the content of `createWith` with the code I posted. That method is the one against I'm primarily running the tests. So the `SET TRANSACTION` wasn't used in my test case last night (either way, thanks for catching that loose call :+1:).

Of course, that method is called from our API, so there is more stuff around it that could explain why I get different results. I was able to reproduce the issue with that code here at work as well, but it's a lot harder. So I was also able to check how 1.9.9 impacts the situation and I'm happy to say that the problem is gone and I now get the `Loose request outside a connection that has expired.` error.

I'm going to try to find out *why* this happens. What's confusing me the most, is that the problem only happens on my work desktop when I put a lot of load on the system. When I run the same code at home, the first query produced the issue instantly.

Thanks for your help! I'll report back when I have any new insights.

**Comment by vitaly-t** [comment]
When event `error` is reported, it has the query as part of the event, so you can see which query exactly causes the problem. Once you find it, you need to track all possible ways in which it is executed, and find the execution path where there is no `.then` following the query, which would constitute a loose promise call. That's basically it.

This is perhaps more of a promise-related issue than anything else.

As for the difference between your home and work setup, it is not surprising because loose promise calls are always dependent on race conditions, and especially in a true asynchronous code.

You see, while V8 engine executes javascript as a single synchronous process, the IO is processed by Node JS asynchronously, which is its strength, and since queries are going through TCP-IP, the otherwise pseudo-synchronous promise requests become truly synchronous, and as such, open to all kinds of race condition conflicts.

**Comment by oliversalzburg** [comment]
So, looking at it some more, it kinda makes sense, but I still don't see how this should be prevented (except for restructuring our queries, which is what I'm going to do).

Looking at the log output from the sample:
```
2015-09-08 10:08:32.469 [DEBUG ] (       postgres.js) tx(tx)/start
2015-09-08 10:08:32.469 [DEBUG ] (       postgres.js) tx(tx): begin
2015-09-08 10:08:32.470 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 10:08:32.470 [DEBUG ] (       postgres.js) tx(tx): SELECT id FROM person WHERE(FALSE)
2015-09-08 10:08:32.470 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 10:08:32.471 [ERROR ] (       postgres.js) tx(tx)/error on database transaction: No data returned from the query.
2015-09-08 10:08:32.471 [INFO  ] (       postgres.js)   SELECT id FROM person WHERE(FALSE)
2015-09-08 10:08:32.477 [DEBUG ] (       postgres.js) tx(tx): rollback
2015-09-08 10:08:32.478 [DEBUG ] (       postgres.js) tx(tx): SELECT id FROM person WHERE(FALSE)
2015-09-08 10:08:32.478 [DEBUG ] (       postgres.js) tx(tx)/endduration: 9, success: false
2015-09-08 10:08:32.481 [DEBUG ] (           pg:pool) timeout: 1441699742481 verbose
2015-09-08 10:08:32.481 [DEBUG ] (           pg:pool) dispense() clients=0 available=2 info
2015-09-08 10:08:32.481 [DEBUG ] (       postgres.js) disconnect(postgres@fairmanager)
2015-09-08 10:08:32.489 [INFO  ] (     apiHandler.js) api(create person)/fail
2015-09-08 10:08:32.489 [ERROR ] (     apiHandler.js) No data returned from the query.
2015-09-08 10:08:32.490 [DEBUG ] (              HTTP) POST /api/people 500 42.691 ms - 34
2015-09-08 10:08:32.490 [ERROR ] (       postgres.js) tx(tx)/error on database transaction: Loose request outside a connection that has expired.
2015-09-08 10:08:32.490 [INFO  ] (       postgres.js)   SELECT id FROM person WHERE(FALSE)

```

I see that the query, which is being detected as a loose call, is `SELECT id FROM person WHERE(FALSE)`. But I don't see how the sample code resulted in that. The promise chain looks perfectly valid to me. Except that there are probably multiple chains, due to the use of `Promise.all` and only the top most promise ultimately rejects when the nested chains completed.

So I'm asking myself if this can actually be prevented and if it needs to be prevented. Especially since it seems like the client reference was still valid when the query was issued, but only became invalid when it's time to handle the resulting error.

Either way, the core issue was the uncaught TypeError, which is now gone. So I can focus on the other issues related to this code :)

**Comment by vitaly-t** [comment]
What I can't understand from your log is the following 3 records:
```
2015-09-08 10:08:32.477 [DEBUG ] (       postgres.js) tx(tx): rollback
2015-09-08 10:08:32.478 [DEBUG ] (       postgres.js) tx(tx): SELECT id FROM person WHERE(FALSE)
2015-09-08 10:08:32.478 [DEBUG ] (       postgres.js) tx(tx)/endduration: 9, success: false
```
It says that the transaction has finished in error, so it is rolled back, and then somehow before the transaction ends it manages to execute a `SELECT`. If you look at the transaction implementation, this is impossible, unless that call is loose, in which case it can still execute at an unpredictable time, i.e. it can happen just before the transaction disconnects or right after - the latter will be caught by the latest code. You must be making a loose call for `SELECT id FROM person` while inside a transaction.

**Comment by oliversalzburg** [comment]
Well, maybe I'm misinterpreting how the code actually works, but this is how it plays out in my mind.

The `rollback` is issued at line 795, but the state of the transaction/client is not changed at that point. The success is determined in another `then` handler. This is handled asynchronously again. So there is an opportunity for other code to run. Only after the `then` handler, will the transaction promise be rejected and future calls would fail.

**Comment by vitaly-t** [comment]
You see, `rollback` can only execute after `invoke()` with all your transaction code has either resolved or rejected. But according to your log, there is a query inside a transaction that hasn't finished execution after `invoke()` has been resolved. That's not possible, unless the request is a loose promise.

**Comment by oliversalzburg** [comment]
Humm, you're probably right. It fits the observed error.

I guess the promise that causes the transaction to be rejected, is the one that is constructed with `Promise.all`. This will be rejected as soon as any contained promise rejects. But that doesn't stop the other contained promises to still be executed.

I adjusted my sample code:
```js
return self.db.tx( function tx( transaction ) {
	return Promise.all( [ foo(), failsFirst() ] )
		.catch( function( error ) {
			log.debug( "Promise.all rejected" );
			throw error;
		} );

	function foo() {
		return succeeds()
			.then( succeeds )
			.then( failsSecond );

		function succeeds() {
			return transaction.one( "SELECT 1 AS id" );
		}
	}

	function failsFirst() {
		return transaction.one( "SELECT id AS first FROM person WHERE(FALSE)" );
	}

	function failsSecond() {
		return transaction.one( "SELECT id AS second FROM person WHERE(FALSE)" );
	}
} );
```

Now the log shows this:
```
2015-09-08 11:03:44.819 [DEBUG ] (       postgres.js) tx(tx)/start
2015-09-08 11:03:44.819 [DEBUG ] (       postgres.js) tx(tx): begin
2015-09-08 11:03:44.820 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 11:03:44.820 [DEBUG ] (       postgres.js) tx(tx): SELECT id AS first FROM person WHERE(FALSE)
2015-09-08 11:03:44.821 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 11:03:44.821 [ERROR ] (       postgres.js) tx(tx)/error on database transaction: No data returned from the query.
2015-09-08 11:03:44.821 [INFO  ] (       postgres.js)   SELECT id AS first FROM person WHERE(FALSE)
2015-09-08 11:03:44.826 [DEBUG ] (  personManager.js) Promise.all rejected
2015-09-08 11:03:44.827 [DEBUG ] (       postgres.js) tx(tx): rollback
2015-09-08 11:03:44.828 [DEBUG ] (       postgres.js) tx(tx): SELECT id AS second FROM person WHERE(FALSE)
2015-09-08 11:03:44.828 [DEBUG ] (       postgres.js) tx(tx)/end duration: 9, success: false
```

**Comment by vitaly-t** [comment]
Why in your function `foo()` do you execute this logic? - if `succeeds()` resolves, then execute `succeed()` again. That's what it does ;)

**Comment by oliversalzburg** [comment]
If I take out one `succeeds()`, the behavior changes. I'm not sure if the length of the promise chain is relevant here or if this just effects the way the resulting queries are interleaved.

**Comment by vitaly-t** [comment]
Does this piece of code show the problem still? It's design suggests that it shouldn't.

**Comment by oliversalzburg** [comment]
Yes, the problem persists. The full log is:
```
2015-09-08 11:03:44.819 [DEBUG ] (       postgres.js) tx(tx)/start
2015-09-08 11:03:44.819 [DEBUG ] (       postgres.js) tx(tx): begin
2015-09-08 11:03:44.820 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 11:03:44.820 [DEBUG ] (       postgres.js) tx(tx): SELECT id AS first FROM person WHERE(FALSE)
2015-09-08 11:03:44.821 [DEBUG ] (       postgres.js) tx(tx): SELECT 1 AS id
2015-09-08 11:03:44.821 [ERROR ] (       postgres.js) tx(tx)/error on database transaction: No data returned from the query.
2015-09-08 11:03:44.821 [INFO  ] (       postgres.js)   SELECT id AS first FROM person WHERE(FALSE)
2015-09-08 11:03:44.826 [DEBUG ] (  personManager.js) Promise.all rejected
2015-09-08 11:03:44.827 [DEBUG ] (       postgres.js) tx(tx): rollback
2015-09-08 11:03:44.828 [DEBUG ] (       postgres.js) tx(tx): SELECT id AS second FROM person WHERE(FALSE)
2015-09-08 11:03:44.828 [DEBUG ] (       postgres.js) tx(tx)/end duration: 9, success: false
2015-09-08 11:03:44.830 [DEBUG ] (           pg:pool) timeout: 1441703054830 verbose
2015-09-08 11:03:44.830 [DEBUG ] (           pg:pool) dispense() clients=0 available=2 info
2015-09-08 11:03:44.830 [DEBUG ] (       postgres.js) disconnect(postgres@fairmanager)
2015-09-08 11:03:44.838 [INFO  ] (     apiHandler.js) api(create person)/fail
2015-09-08 11:03:44.838 [ERROR ] (     apiHandler.js) No data returned from the query.
2015-09-08 11:03:44.840 [DEBUG ] (              HTTP) POST /api/people 500 39.770 ms - 34
2015-09-08 11:03:44.840 [ERROR ] (       postgres.js) tx(tx)/error on database transaction: Loose request outside a connection that has expired.
2015-09-08 11:03:44.840 [INFO  ] (       postgres.js)   SELECT id AS second FROM person WHERE(FALSE)
```

**Comment by vitaly-t** [comment]
I'm still unable to reproduce the issue on my PC. Could be a result of a completely different configuration, impossible to know.

I'm on NodeJS 0.12.7, 64-bit, on Windows 10.

I remotely understand the issue, which looks like `failsSecond` is being resolved after the transaction execution, according to your log. But I can't reproduce it here, it always just works correctly.

I will keep trying to break it :)

**Comment by vitaly-t** [comment]
Ok, I've been able to reproduce it, and I can see what the problem is now...

`Promise.all` stops resolving the chain returned by function `foo` when the promise returned by `failsFirst` has rejected. The promise chain continues to be resolved however, after `promise.all` has rejected and the transaction has ended.

It became possible to see when I changed the code as follows:

```javascript
    function foo() {
        return succeeds()
            .then(succeeds)
            .then(function () {
                return new promise.Promise(function (resolve, reject) {
                    setTimeout(function () {
                        resolve();
                    });
                });
            })
            .then(failsSecond);

        function succeeds() {
            return transaction.one( "SELECT 1 AS id" );
        }
    }
```

**Comment by oliversalzburg** [comment]
Very interesting! I had also noticed that the problem went away when using `Promise.settle` instead of `Promise.all`. Obviously, because it waits for all promises to be resolved/rejected. I just didn't go for it because I wasn't sure what else it implies in this context.

In case it still matters, my work workstation is Windows 7 x64 Node 0.12.7. My home workstation is Windows 10 x64 Node 0.12.7.

**Comment by vitaly-t** [comment]
I've just looked up [method settle](https://github.com/petkaantonov/bluebird/blob/master/API.md#settle---promise), and it looks like the way to go indeed for this situation. I just couldn't advise it in general documentation, because that method is outside the generic promise protocol.

However, tell me if I'm wrong, but the way library works since 1.9.9, everything works, except you are getting an error logged that should be ignored - correct?

**Comment by oliversalzburg** [comment]
Yes, I guess, everything is working as it should or as well as it possibly can. The problem is now that our code is producing this situation, but I'm working on resolving that :)

**Comment by vitaly-t** [comment]
Basically, what needs to be changed - do not use promise chains as parameters for `promise.all` inside a transaction, as those may result in loose query requests.

In your case, you can just chain the entire transaction explicitly, without using `promise.all` at all.

Or, you can just use `settle` instead :)

Back when I was looking first at the issue with loose requests in large transactions I implemented method `sequence` (see [Synchronous Transactions](https://github.com/vitaly-t/pg-promise#synchronous-transactions)) that's available inside tasks and transactions, which takes care of loose queries.

I will contemplate on making changes in the library to cater for the loose requests better.

**Comment by fundon** [comment]
see [bluebird.reduce](https://github.com/petkaantonov/bluebird/blob/master/API.md#reducefunction-reducer--dynamic-initialvalue---promise)

**Comment by vitaly-t** [comment]
Yeah, I know it quite well, but I needed an implementation that would work with any standard promise library, while `reduce` is outside the promise standard.

**Comment by oliversalzburg** [comment]
I had actually looked at `sequence` but decided that I wouldn't need it, because I *do* want parallelization. But I'll rethink this once more :)

Thanks for the help!

**Comment by vitaly-t** [comment]
Cheers! This has been very useful. And the library now can anticipate loose requests and reject them correctly. I don't know if there is anything else it can or should do in such cases, but I will think about it. Feel free throwing in your own ideas regarding it.

In the meantime, the issue has been resolved, it seems. Closing it.

**Comment by vitaly-t** [comment]
@oliversalzburg 

Playing around with `promise.all`a bit more, I found even more disturbing behaviour. Sometimes a loose change request ('insert' or `update`) would apply the change when executed after the `rollback`, which makes this a more dangerous premise. Makes me question how much is `promise.all` really suitable to transactions in general then.

`settle` on the other hand guarantees a consistent transaction, so does `sequence`, but they have they own issues: `settle` is proprietary, exists only in Bluebird, while `sequence` was designed for something different (massive transactions), and isn't as easy to use.

Right now I'm thinking... maybe I should extend the library with a method similar to `settle`, just easier to use, so it would work the same with any promise library.

What do you reckon?

**Comment by vitaly-t** [comment]
I have done lots of research on the subject of a proper promise implementation for transactions, and have come up with the first draft that works exactly in the way a transaction needs:

```javascript
// Joins `all` + `settled` logic for transactions;
function allSettled(inputs) {
    if (!inputs.length) {
        return promise.resolve([]);
    }
    return new promise.Promise(function (resolve, reject) {
        var error, success = true, remaining = inputs.length,
            result = new Array(remaining);
        inputs.forEach(function (item, i) {
            item
                .then(function (data) {
                    result[i] = data;
                    check();
                }, function (reason) {
                    if (success) {
                        error = reason;
                        success = false;
                    }
                    check();
                });
        });
        function check() {
            if (!--remaining) {
                if (success) {
                    resolve(result);
                } else {
                    reject(error);
                }
            }
        }
    });
}
```
It will take a few tests and some documentation before it is ready for release, but now I'm confident that this is the way to go.

Also, need to figure out the best name for such a method Any suggestions? :)

**Comment by oliversalzburg** [comment]
@vitaly-t Wow, excellent work! I'll try to test this soon, but I have to work on progressing our project further right now. I spent a lot of time on load and concurrency issues the past few days and people are waiting for a merge :P

**Comment by vitaly-t** [comment]
That's alright, I'm throwing in some good tests at it now, and finalizing the protocol changes. Will try to do a new release soon with all the new functionality tested and documented.

**Comment by vitaly-t** [comment]
@oliversalzburg I took my time to do it right, to thoroughly test, optimize and document, but now the [new release](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.10.0) is out.

It is a major improvement on top of what was initially contemplated. And while the new method `batch` resolves with the same results as `promise.all`, which makes it easy to replace the old calls, it rejects differently - see [the release notes](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.10.0).

I appreciate if you can find time to try it and come back with a feedback here.

**Comment by oliversalzburg** [comment]
I'll definitely upgrade ASAP, but I'll be on a short trip until the end of the week. I'll report back next week :)

**Comment by vitaly-t** [comment]
By the way, I've decided on a new project - [spe](https://github.com/vitaly-t/spe) where I'm moving things like `sequence` and `batch` methods, and will be adding more later. They just seem to be too genetic not to :) and quite useful too.

**Comment by vitaly-t** [comment]
@oliversalzburg You never came back to me about the use of method `batch`. Have you changed your code to use it?

**Comment by oliversalzburg** [comment]
@vitaly-t Not yet. This is still on my radar though :) In fact, I'll switch out our `.all()` calls right now.

**Comment by oliversalzburg** [comment]
All tests are passing and I can see the new failing behavior in edge cases. I'll just have to adjust our error handling because it doesn't expect the resolution array we're getting now.

```
2015-09-16 09:49:10.000 [ERROR ] (     apiHandler.js) [
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": true
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": true
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": false,
2015-09-16 09:49:10.000                                   "result": "Single row was expected from the query, but multiple returned."
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": false,
2015-09-16 09:49:10.000                                   "result": "Single row was expected from the query, but multiple returned."
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": false,
2015-09-16 09:49:10.000                                   "result": "Single row was expected from the query, but multiple returned."
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": false,
2015-09-16 09:49:10.000                                   "result": "Single row was expected from the query, but multiple returned."
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": false,
2015-09-16 09:49:10.000                                   "result": "Single row was expected from the query, but multiple returned."
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": true
2015-09-16 09:49:10.000                                 },
2015-09-16 09:49:10.000                                 {
2015-09-16 09:49:10.000                                   "success": true
2015-09-16 09:49:10.000                                 }
2015-09-16 09:49:10.000                               ]
```

**Comment by vitaly-t** [comment]
Ok, here's the [full batch API](https://github.com/vitaly-t/pg-promise/blob/master/API.md#module_pg-promise.Task+batch). Please keep me updated on how you're getting along with it ;)

**Comment by oliversalzburg** [comment]
Nice catch! Yeah, we're using `JSON.stringify` there. Should be [fixed](https://github.com/hartwig-at/fm-log/commit/a698b240552ad998034dc5077fb80d5d11214478) now (not optimal, but it'll do for now).

We're on 1.10.5 since yesterday :)

**Comment by vitaly-t** [comment]
I probably should have added [this kind of example](https://github.com/vitaly-t/pg-promise/releases/tag/v.1.10.6) a while ago :)

**Comment by vitaly-t** [comment]
@oliversalzburg if you are still suing this library...

This may be of interest to you: [Performance Boost](https://github.com/vitaly-t/pg-promise/wiki/Performance-Boost) ;)

**Comment by oliversalzburg** [comment]
@vitaly-t Sure, we're still using it ;) I'll check out the article in more detail later. But I don't think it's relevant for us at this time.
